File: common/room/compose/src/androidMain/kotlin/modal/add_room/AddRoomModal.kt
package modal.add_room

import androidx.compose.runtime.Composable
import io.github.alexgladkov.kviewmodel.compose.ViewModel
import io.github.alexgladkov.kviewmodel.compose.observeAsState
import io.github.alexgladkov.kviewmodel.odyssey.StoredViewModel
import modal.add_room.models.AddRoomAction
import modal.add_room.models.AddRoomEvent
import modal.add_room.view.DoneView
import modal.add_room.view.RoomNameView
import modal.add_room.view.SelectDevicesView
import modal.add_room.view.SelectImageView
import ru.alexgladkov.odyssey.compose.local.LocalRootController

@Composable
fun AddRoomModal(onCloseClick:() -> Unit) {

    ViewModel(factory = { AddRoomViewModel() }){ viewModel ->
        val state = viewModel.viewStates().observeAsState()
        val action = viewModel.viewActions().observeAsState()


        when(action.value){

            is AddRoomAction.OpenRoomName -> {
                RoomNameView(viewState = state.value){
                    viewModel.obtainEvent(it)
                }
            }

            is AddRoomAction.OpenSelectImage -> {
                SelectImageView(viewState = state.value){
                    viewModel.obtainEvent(it)
                }
            }

            is AddRoomAction.OpenSelectDevices -> {
                SelectDevicesView(viewState = state.value){
                    viewModel.obtainEvent(it)
                }
            }

            is AddRoomAction.OpenDone -> {
                DoneView(){
                    viewModel.obtainEvent(it)
                }
            }

            is AddRoomAction.OpenRoomScreen -> {
                onCloseClick()
            }

            else -> {}
        }

    }
}

File: iosApp/iosApp/Screens/AddRoom/AddRoomModal.swift
//
//  AddRoomScreen.swift
//  iosApp
//
//  Created by MacBook on 19.02.2024.
//  Copyright © 2024 orgName. All rights reserved.
//

import SwiftUI
import SharedSDK

struct AddRoomModal: View {
    
    let onClose:() -> Void
    private let viewModel = AddRoomViewModel()
    @State private var currentView:AddRoomAction? = nil
    
    
    
    var body: some View {
        
        let _ = print(currentView)
        
        ObservingView(statePublisher: statePublisher(viewModel.viewStates())){ viewState in
            
            switch(currentView){
                case AddRoomAction.OpenRoomName():
                    RoomNameView(viewState: viewState){event in
                        viewModel.obtainEvent(viewEvent: event)
                    }
                case AddRoomAction.OpenSelectImage():
                    SelectImageView(viewState: viewState){event in
                        viewModel.obtainEvent(viewEvent: event)
                    }
                case AddRoomAction.OpenSelectDevices():
                    SelectDevicesView(viewState: viewState){event in
                        viewModel.obtainEvent(viewEvent: event)
                    }
                case AddRoomAction.OpenDone():
                    DoneView(viewState: viewState){event in
                        viewModel.obtainEvent(viewEvent: event)
                    }
                case AddRoomAction.OpenRoomScreen():
                    let _ = onClose()
               
                default:
                    EmptyView()
            }
            
        }
        .onReceive(sharePublisher(viewModel.viewActions())){ action in
            currentView = action
        }
    }
}

